Inicjalizacja zadań Grupy 1 - Teren

rafa1867 opened the pull request.

rafa1867 committed:
Inicjalizacja zadań Grupy 1 - Teren

rafa1867 committed:
Dodane tekstury lasu, trawy, betonu, cmentarza. dodany efekt wiatru dla tekstur lasu i trawy oraz dodany efekt 3d dla cmentarzy

shajder commented:
Please complete below steps after your next commit:

-sync your fork 
-merge main branch into your task-teren branch
-git push
-refresh your map_data folder with updated landuse shapefile

If you will face any problem with it just feel free to let me know, thanks!

Wktr-24 committed:
feat: add missing landuse types and textures

Implement handling for scrub, farmland, allotments, etc.
Add optimized textures for new terrain types.

Wktr-24 committed:
fix: resolve Z-fighting with PolygonOffset layer system

shajder commented on src/landuse.cpp:
```diff
@@ -1,112 +1,363 @@
-    // draw terrain first
-    land_model->getOrCreateStateSet()->setRenderBinDetails(-10, "RenderBin");
-    // do not nest this render bin
-    land_model->getOrCreateStateSet()->setNestRenderBins(false);
```
Think of a terrain as a first layer to be drawn - background. Thus it does not need to set z-value, as result you will avoid z-fighting issue. This is why I passed false for zWrite parametere of osg::Depth function.

shajder commented:
-You currently have a conflict in src/landuse.cpp. This means you’ll need to resolve it, which is not very complex. Below are the steps to sort it out:
a) sync your fork to fetch current origin/main
b) `git fetch origin`
c) `git chceckout `[name of your working branch]
d) `git merge origin/main` // here you will get an information about the conflict, stay calm
e) In the conflicting file (src/landuse.cpp), look for the section marked HEAD ... main. This is how Git shows the two conflicting versions of the code.
f) Decide which code is correct:
the HEAD section contains your changes
the main section represents the code currently in the main branch
g) Fix the code and build the project to make sure all conflicts are resolved
h) `git add` [list of conflicting files you resolved]
i) `git commit -m "Resolved conflicts or so"`
j) `git push`

Wktr-24 committed:
fix: replace PolygonOffset with render bin configuration

Disable Z-write for terrain and set render bin to -10 to resolve Z-fighting issues.

rafa1867 committed:
Dodanie logiki cashe first

Wktr-24 committed:
refactor: DDS textures, remove GIS binaries

michalmer committed:
Merge branch 'main' into task-teren

shajder commented on data/shaders/parallax.frag:
```diff
@@ -0,0 +1,37 @@
+uniform sampler2D baseTexture;
```
All shaders put to `data/shaders` directory

shajder commented on src/landuse.cpp:
```diff
@@ -1,100 +1,384 @@
+        texture->setWrap(osg::Texture::WRAP_T, osg::Texture::REPEAT);
+
+        texture->setFilter(osg::Texture::MIN_FILTER, osg::Texture::LINEAR);
+        texture->setFilter(osg::Texture::MAG_FILTER, osg::Texture::LINEAR);
```
```suggestion
        texture->setFilter(osg::Texture::MIN_FILTER, osg::Texture::LINEAR_MIPMAP_LINEAR);
        texture->setFilter(osg::Texture::MAG_FILTER, osg::Texture::LINEAR);
        texture->setUseHardwareMipMapGeneration(true);
```
Switch on mipmap filtering

shajder commented on src/landuse.cpp:
```diff
@@ -1,100 +1,384 @@
+    osg::ref_ptr<osg::Shader> vert = osgDB::readRefShaderFile(
+        osg::Shader::VERTEX, data_path + "/standard.vert");
+    osg::ref_ptr<osg::Shader> frag = osgDB::readRefShaderFile(
+        osg::Shader::FRAGMENT, data_path + "/standard.frag");
```
```suggestion
    osg::ref_ptr<osg::Shader> vert = osgDB::readRefShaderFile(
        osg::Shader::VERTEX, "shaders/standard.vert");
    osg::ref_ptr<osg::Shader> frag = osgDB::readRefShaderFile(
        osg::Shader::FRAGMENT, "shaders/standard.frag");
```
Assume data is working directory (data_path not needed), put shaders into shaders directory.

shajder commented on src/landuse.cpp:
```diff
@@ -1,100 +1,384 @@
+    osg::ref_ptr<osg::Shader> vert =
+        osgDB::readRefShaderFile(osg::Shader::VERTEX, data_path + "/wind.vert");
+    osg::ref_ptr<osg::Shader> frag = osgDB::readRefShaderFile(
+        osg::Shader::FRAGMENT, data_path + "/wind.frag");
```
data_path not needed here as well

shajder commented on src/landuse.cpp:
```diff
@@ -1,100 +1,384 @@
+    osg::ref_ptr<osg::Shader> vert = osgDB::readRefShaderFile(
+        osg::Shader::VERTEX, data_path + "/parallax.vert");
+    osg::ref_ptr<osg::Shader> frag = osgDB::readRefShaderFile(
+        osg::Shader::FRAGMENT, data_path + "/parallax.frag");
```
same here, data_path not needed here

shajder commented on src/landuse.cpp:
```diff
@@ -1,100 +1,384 @@
+        h_tex->setWrap(osg::Texture::WRAP_S, osg::Texture::REPEAT);
+        h_tex->setWrap(osg::Texture::WRAP_T, osg::Texture::REPEAT);
+        h_tex->setFilter(osg::Texture::MIN_FILTER, osg::Texture::LINEAR);
+        h_tex->setFilter(osg::Texture::MAG_FILTER, osg::Texture::LINEAR);
```
```suggestion
        h_tex->setFilter(osg::Texture::MIN_FILTER, osg::Texture::LINEAR_MIPMAP_LINEAR);
        h_tex->setFilter(osg::Texture::MAG_FILTER, osg::Texture::LINEAR);
        h_tex->setUseHardwareMipMapGeneration(true);
```

I guess you need mipmap filtering here as well

shajder commented on src/landuse.cpp:
```diff
@@ -1,100 +1,384 @@
+        else if (name.find("cemetery") != std::string::npos)
+        {
+            apply_texture(ss, file_path + "/images/cemetery.dds");
+            setup_parallax_shader(ss, file_path);
```
Eventually, you do not need to use `file_path `varaible in this section.

shajder reviewed (changes requested):
In general, it looks good. Few minor issues to correct:

-put shaders into data/shaders directory
-assume `data `as working directory, data path is only for shapefiles
-generate texture mipmaps and use mipmap minifying filter to get rid of aliasing

Wktr-24 committed:
fix: shaders to data/shaders, mipmaps, paths

Wktr-24 committed:
Merge branch 'main' into task-teren

shajder reviewed (changes requested):
Great work on getting those previous issues sorted out! The progress on the branch is looking solid.

However, I’ve just spotted a regression that seems to have been introduced with the latest PR. Here are the details:

The HUD text in the top-left corner is no longer displaying the correct intersection names. If you check the current main branch, the intersection names appear as expected. In the current PR branch, this functionality appears to be broken or misaligned.

At the moment I have no time to search for the problem. My intuition says this could be one of two problems:

1) removed meta-data from the scene. If you need to create new geometry just add user data
2) incorrect resolution of the conflicts

I will take a look in the evening, you have time to figure out by yourself until that.

shajder commented on data/shaders/wind.vert:
```diff
@@ -0,0 +1,11 @@
+void main() {
+    texCoord = gl_Vertex.xy * 0.05;
+    float speed = 0.8; 
+    float strength = 0.02;
```
```suggestion
uniform float texCoordScale;
uniform float animStrength;
uniform float animSpeed;

void main() {
    texCoord = gl_Vertex.xy * texCoordScale;
    float speed = animSpeed; 
    float strength = animStrength;
```

Could we have those parameters passed from command line to uniforms to have control over wind animation?

shajder replied:
Oh, and one more thing. Use those uniforms for standard and paralax shaders as well.

shajder commented on src/landuse.cpp:
```diff
@@ -1,100 +1,388 @@
+}
+
+// Shader dla cmentarza (Parallax 3D)
+void setup_parallax_shader(osg::StateSet* ss)
```
I bet you noticed something is off here, I will take a closer look later but feel free to deliver some intel or thoughts:

![screen_shot_0_0](https://github.com/user-attachments/assets/b0fbe3c3-ee59-408a-9cdc-8012b5df96f7)

shajder replied:
I found the problem. In the same location there are two types of geometry:

1) one you identify as cementary
2) something else

When bounding box of "something else" is at the edge of viewing frustum its geometry became visible and we see white thingy. This also applies to other geometries. We need to remove all (2) geometies. I applied correction to the main loop of your code:


```
    for (Mapping::iterator it = umap.begin(); it != umap.end(); ++it)
    {
        std::string name = it->first;
        std::vector<osg::ref_ptr<osg::Node>>& nodes = it->second;
        osg::ref_ptr<osg::StateSet> ss = new osg::StateSet();

        // === WARSTWA 0-3: TERENY MIEJSKIE (na spodzie) ===
        if (name.find("residential") != std::string::npos)
        {
            apply_texture(ss, "images/concrete.dds");
            setup_standard_shader(ss);
        }
        else if (name.find("industrial") != std::string::npos)
        {
            apply_texture(ss, "images/concrete.dds");
            setup_standard_shader(ss);
        }
        else if (name.find("commercial") != std::string::npos)
        {
            apply_texture(ss, "images/concrete.dds");
            setup_standard_shader(ss);
        }
        else if (name.find("retail") != std::string::npos)
        {
            apply_texture(ss, "images/concrete.dds");
            setup_standard_shader(ss);
        }
        // === WARSTWA 4-6: ROLNICTWO, WOJSKO, KAMIENIOŁOMY ===
        else if (name.find("farmland") != std::string::npos)
        {
            apply_texture(ss, "images/farmland.dds");
            setup_standard_shader(ss);
        }
        else if (name.find("farmyard") != std::string::npos)
        {
            apply_texture(ss, "images/farmland.dds");
            setup_standard_shader(ss);
        }
        else if (name.find("quarry") != std::string::npos)
        {
            apply_texture(ss, "images/rock.dds");
            setup_standard_shader(ss);
        }
        else if (name.find("military") != std::string::npos)
        {
            apply_texture(ss, "images/military.dds");
            setup_standard_shader(ss);
        }
        // === WARSTWA 8-10: TRAWA, ŁĄKI, ZAROŚLA ===
        else if (name.find("grass") != std::string::npos)
        {
            apply_texture(ss, "images/grass.dds");
            setup_wind_shader(ss);
        }
        else if (name.find("meadow") != std::string::npos)
        {
            apply_texture(ss, "images/grass.dds");
            setup_wind_shader(ss);
        }
        else if (name.find("scrub") != std::string::npos)
        {
            apply_texture(ss, "images/scrub.dds");
            setup_wind_shader(ss);
        }
        else if (name.find("heath") != std::string::npos)
        {
            apply_texture(ss, "images/scrub.dds");
            setup_wind_shader(ss);
        }
        // === WARSTWA 12: LASY ===
        else if (name.find("forest") != std::string::npos)
        {
            apply_texture(ss, "images/forest.dds");
            setup_wind_shader(ss);
        }
        // === WARSTWA 13-15: SADY, DZIAŁKI, REZERWATY ===
        else if (name.find("orchard") != std::string::npos)
        {
            apply_texture(ss, "images/orchard.dds");
            setup_wind_shader(ss);
        }
        else if (name.find("nature_reserve") != std::string::npos)
        {
            apply_texture(ss, "images/orchard.dds");
            setup_wind_shader(ss);
        }
        else if (name.find("allotments") != std::string::npos)
        {
            apply_texture(ss, "images/allotments.dds");
            setup_wind_shader(ss);
        }
        // === WARSTWA 16-17: PARKI, TERENY REKREACYJNE ===
        else if (name.find("park") != std::string::npos)
        {
            apply_texture(ss, "images/grass.dds");
            setup_wind_shader(ss);
        }
        else if (name.find("recreation_ground") != std::string::npos)
        {
            apply_texture(ss, "images/sport_green.dds");
            setup_wind_shader(ss);
        }
        // === WARSTWA 18: CMENTARZ (najwyżej - parallax) ===
        else if (name.find("cemetery") != std::string::npos)
        {
            apply_texture(ss, "images/cemetery.dds");
            setup_parallax_shader(ss);
        }
        else
        {
            for (size_t i = 0; i < nodes.size(); ++i)
            {
                while ( nodes[i].valid() && nodes[i]->getNumParents() )
                    nodes[i]->getParent(nodes[i]->getNumParents()-1)->removeChild(nodes[i]);
            }
            continue;
        }
        for (size_t i = 0; i < nodes.size(); ++i)
        {
            if (nodes[i].valid()) nodes[i]->setStateSet(ss);
        }
    }

```

shajder reviewed (changes requested):
I found the problem, something went wront with merging main repository. ~~In map.cpp, line 432, you should have this: getLandInfoAtIntersection(scene, hit);~~

~~scene instead root~~

EDIT: now I had a bit more time to take a closer look, forget about HUD problem, just focus on the problems listed below.

Wktr-24 committed:
feat: add shader uniforms and remove unrecognized geometries

shajder reviewed (approved):
Please correct clang format and we are ready to merge

shajder commented:
> Please correct clang format and we are ready to merge

@rafa1867 could you fix CI/CD check for clang format? 2 lines of code I think:

<img width="1479" height="152" alt="image" src="https://github.com/user-attachments/assets/dc03baeb-3ca0-4fce-8de0-c33e251f7ad7" />

Wktr-24 committed:
style: fix clang-format

shajder reviewed (approved):
Okay, I am going to approve and merge this. However, you should be aware that one important part of the project is still missing.

According to the project description, lighting support should be implemented. For effects such as parallax mapping or normal mapping, this should be done using per-pixel lighting.

I understand that you likely don’t have time to complete this at the moment, so we will proceed with the current version as it is.
